### PR TITLE
0250 Next RPC Indication for the HMI

### PR DESCRIPTION
**Pull request by MichaelCrimando**
In situations with driver distraction, also with making a good UI, it really hurts the possible UX when the HMI has no idea what's going to happen next after the user taps an in-app menu item, a softbutton or a selection in a perform interaction. This feature would provide apps with a way for the HMI to know what's next when a button or choice is selected.

This can be used in tandem with Template Improvements: Additional SubMenus to provide a fantastic user experience.

### CODE REVIEW

**Comment by MichaelCrimando** [comment]
Setting this PR up now for when the proposal gets accepted with revisions

**Comment by MichaelCrimando** [comment]
@joeljfischer I set up this PR for when the original proposal gets accepted with revisions

Does the ObjC section make sense?

File: proposals/0250-NextRpcIndication.md [hunk 1]
```diff
@@ -96,6 +96,88 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
 </struct>
  ```
 
+### iOS
+1. We're adding in a public manager-level enum
+(iOS example in Swift that we should translate to ObjC)
+```ObjectiveC
+typedef NS_ENUM(NSInteger, NextFunction) {
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]:
```suggestion
typedef NS_ENUM(NSInteger, SDLNextFunction) {
```

The obj-c library needs to prefix things with `SDL` for namespacing

File: proposals/0250-NextRpcIndication.md [hunk 2]
```diff
@@ -96,6 +96,88 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+        Default =0,
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]: All of these need to be prefixed with `SDLNextFunction` on Obj-C

File: proposals/0250-NextRpcIndication.md [hunk 3]
```diff
@@ -96,6 +96,88 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+        AccessMicrophone, 
+        ScrollableMessage, 
+        Slider, 
+        SendLocation, 
+        DialNumber, 
+        OpenMenu 
+};
+```
+
+2. Add a manager-level class `NextFunctionInfo` with an initializer
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]: This departs from the design we talked about by adding a second class that mirrors and mimics the existing RPC class almost exactly. This kind of code-duplication is unnecessary. You can just use the NextFunctionInfo RPC class and add an initializer to it.

File: proposals/0250-NextRpcIndication.md [hunk 4]
```diff
@@ -96,6 +96,88 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+```ObjectiveC
+@interface NextFunctionInfo: NSObject 
+	@property (copy, nonatomic, readonly) NextFunction *nextFunction;
+	@property (copy, nonatomic, readonly, nullable) NSString *loadingText;
+@end
+```
+
+NextFunctionInfo.m
+```ObjectiveC
+@implementation NextFunctionInfo
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]: You don't need to show implementation code in the proposal

File: proposals/0250-NextRpcIndication.md [hunk 5]
```diff
@@ -96,6 +96,88 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+        AccessMicrophone, 
+        ScrollableMessage, 
+        Slider, 
+        SendLocation, 
+        DialNumber, 
+        OpenMenu 
+}
+```
+
+2. Add a manager-level class `NextFunctionInfo` with an initializer
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]: See above

File: proposals/0250-NextRpcIndication.md (hunk 3 above)
Comment (MichaelCrimando) [reply proposals/0250-NextRpcIndication.md]: Hmm I dont really understand how to modify what I have into what you're talking about here

Comment (MichaelCrimando) [reply proposals/0250-NextRpcIndication.md]: @joeljfischer plz help

Comment (joeljfischer) [reply proposals/0250-NextRpcIndication.md]: The design we talked about was simply to add an initializer to the RPC class that takes an `SDLNextFunction` and converts it into the function ID. Your design here is copying the RPC class into a "high-level manager class" with minimal changes. That seems to me to be code duplication without a good reason.

File: proposals/0250-NextRpcIndication.md [hunk 6]
```diff
@@ -96,6 +96,87 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+```ObjectiveC
+@interface NextFunctionInfo: NSObject 
+	@property (copy, nonatomic, readonly) SDLNextFunction *nextFunction;
+	@property (copy, nonatomic, readonly, nullable) NSString *loadingText;
+@end
+```
+
+NextFunctionInfo.m
+```ObjectiveC
+@implementation NextFunctionInfo
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]: This implementation code is unnecessary. Only non-obvious or implementation code that could be implemented in several different ways needs to be added.

**Comment by MichaelCrimando** [comment]
@joeljfischer I pushed in an update - does this look better?

File: proposals/0250-NextRpcIndication.md [hunk 7]
```diff
@@ -96,6 +96,48 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+        SDLNextFunctionScrollableMessage, 
+        SDLNextFunctionSlider, 
+        SDLNextFunctionSendLocation, 
+        SDLNextFunctionDialNumber, 
+        SDLNextFunctionOpenMenu 
+};
+```
+
+2. Add an initializer to all RPC classes that take `SDLNextFunction` and converts it into the functionID.
+3. Add a `nextFunctionInfo` parameter to `SDLChoiceCell`, `SDLMenuCell`, and `SDLSoftButtonObject` and their respective initilaizers.
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]:
```suggestion
3. Add a `nextFunctionInfo` parameter to `SDLChoiceCell`, `SDLMenuCell`, and `SDLSoftButtonObject` and their respective initializers. 
```

File: proposals/0250-NextRpcIndication.md [hunk 8]
```diff
@@ -96,6 +96,48 @@ Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout
...
+        ScrollableMessage, 
+        Slider, 
+        SendLocation, 
+        DialNumber, 
+        OpenMenu 
+}
+```
+
+2. Add an initializer to all RPC classes that take `SDLNextFunction` and converts it into the functionID.
+3. Add a `nextFunctionInfo` parameter to `SDLChoiceCell`, `SDLMenuCell`, and `SDLSoftButtonObject` and their respective initilaizers.
```
Comment (joeljfischer) [inline proposals/0250-NextRpcIndication.md]:
```suggestion
3. Add a `nextFunctionInfo` parameter to `SDLChoiceCell`, `SDLMenuCell`, and `SDLSoftButtonObject` and their respective initializers.
```

**Comment by MichaelCrimando** [comment]
ok @joeljfischer 
I fixed your two typos! It should be good to go now

**Comment by MichaelCrimando** [comment]
Thanks!
@theresalech  this is ready to go

**Pull request merged**